android/translate-showcase/app/src/main/java/com/google/mlkit/showcase/translate/main/MainFragment.kt

```diff
@@ -140,77 +192,73 @@ class MainFragment : Fragment() {
+            // Build and bind the camera use cases
+            bindCameraUseCases()
+        }, ContextCompat.getMainExecutor(requireContext()))
+    }
 
-            // To update the SurfaceTexture, we have to remove it and re-add it
-            val parent = viewfinder.parent as ViewGroup
-            parent.removeView(viewfinder)
-            parent.addView(viewfinder, 0)
+    private fun bindCameraUseCases() {
+        val cameraProvider = cameraProvider
+            ?: throw IllegalStateException("Camera initialization failed.")
 
-            viewfinder.surfaceTexture = it.surfaceTexture
-            updateTransform()
-        }
+        // Get screen metrics used to setup camera for full screen resolution
+        val metrics = DisplayMetrics().also { viewFinder.display.getRealMetrics(it) }
+        Log.d(TAG, "Screen metrics: ${metrics.widthPixels} x ${metrics.heightPixels}")
 
-        // Setup image analysis pipeline that computes average pixel luminance
-        val analyzerConfig = ImageAnalysisConfig.Builder().apply {
-            // Use a worker thread for image analysis to prevent glitches
-            val analyzerThread = HandlerThread(
-                "TextAnalysis"
-            ).apply { start() }
-            setCallbackHandler(Handler(analyzerThread.looper))
-            // In our analysis, we care more about the latest image than
-            // analyzing *every* image
-            setImageReaderMode(
-                ImageAnalysis.ImageReaderMode.ACQUIRE_LATEST_IMAGE
-            )
-        }.build()
+        val screenAspectRatio = aspectRatio(metrics.widthPixels, metrics.heightPixels)
+        Log.d(TAG, "Preview aspect ratio: $screenAspectRatio")
+
+        val rotation = viewFinder.display.rotation
+
+        val preview = Preview.Builder()
+            .setTargetAspectRatio(screenAspectRatio)
+            .setTargetRotation(rotation)
+            .build()
 
         // Build the image analysis use case and instantiate our analyzer
-        viewModel.sourceText.observe(viewLifecycleOwner, Observer { srcText.text = it })
-        val analyzerUseCase = ImageAnalysis(analyzerConfig).apply {
-            analyzer =
-                TextAnalyzer(
-                    requireContext(),
-                    viewModel.sourceText,
-                    widthCropPercent = WIDTH_CROP_PERCENT,
-                    heightCropPercent = HEIGHT_CROP_PERCENT
+        imageAnalyzer = ImageAnalysis.Builder()
+            // We request aspect ratio but no resolution
+            .setTargetAspectRatio(screenAspectRatio)
+            .setTargetRotation(rotation)
+            .build()
+            .also {
+                it.setAnalyzer(
+                    cameraExecutor
+                    , TextAnalyzer(
+                        requireContext(),
+                        viewModel.sourceText,
+                        widthCropPercent = WIDTH_CROP_PERCENT,
+                        heightCropPercent = HEIGHT_CROP_PERCENT
```

Comment:
I think this might be problematic in some devices. When you request an aspect ratio, it's not guaranteed you'll get that exact ratio depending on what's available from the camera stack as well as what other use cases are running simultaneously. If you are going to manipulate the images, you should determine things like this *after* you know what size the images coming from the camera have.

Currently, the only way to know the size of the images that will come to the analyzer is to wait until the first image comes (which is less than ideal, unfortunately). So I would suggest moving any logic about passing manipulation to the inside of the analyzer function and instead pass something closer to your "intent" as a parameter. E.g. if you want a squared image, provide an aspect ratio and then compute the appropriate cropping necessary inside of the analyze function.

Reply:
The size of the image coming from the analyzer is not a guarantee. You are asking for a specific aspect ratio, but it's only a "request" -- you could get *anything* back from the camera, unfortunately. You are pre-selecting height and width crop percent, before you know what size / aspect ratio the image will have.

Let me pick a hypothetical (but absurd) example. If a device could only return images to the analyzer with resolution 100x900 (very, very tall and skinny images) then will your logic for cropping still work? If I understand your cropping logic correctly, I suspect there's something else you can do in cases where the image aspect ratio does not match what you requested.

Reply:
Ahh, I think I see the problem now, will modify the logic. Thanks for explaining it again.